Mathlib/Combinatorics/Digraph/Basic.lean

```diff
@@ -0,0 +1,235 @@
+Note that every vertex is adjacent to itself in `⊤`.
+-/
+protected def completeDigraph (V : Type*) : Digraph V where Adj := ⊤
+
+/--
+The empty digraph on a type `V` (denoted by `⊥`)
+is the digraph such that no pairs of vertices are adjacent.
+Note that `⊥` is called the empty digraph because it has no edges.
+-/
+protected def emptyDigraph (V : Type*) : Digraph V where Adj _ _ := False
+
+/--
+Two vertices are adjacent in the complete bipartite digraph on two vertex types
+if and only if they are not from the same side.
+Any bipartite digraph may be regarded as a subgraph of one of these.
+-/
+@[simps]
+def completeBipartiteGraph (V W : Type*) : Digraph (Sum V W) where
+  Adj v w := v.isLeft ∧ w.isRight ∨ v.isRight ∧ w.isLeft
+
+variable {ι : Sort*} {V W X : Type*} (G : Digraph V) (G' : Digraph W) {a b c u v w : V}
+
+theorem adj_injective : Injective (Adj : Digraph V → V → V → Prop) := fun _ _ ↦ Digraph.ext
+
+@[simp] theorem adj_inj {G H : Digraph V} : G.Adj = H.Adj ↔ G = H := Digraph.ext_iff.symm
+
+section Order
+
+/--
+The relation that one `Digraph` is a spanning subgraph of another.
+Note that `Digraph.IsSubgraph G H` should be spelled `G ≤ H`.
+-/
+protected def IsSubgraph (x y : Digraph V) : Prop :=
+  ∀ ⦃v w : V⦄, x.Adj v w → y.Adj v w
+
+instance : LE (Digraph V) := ⟨Digraph.IsSubgraph⟩
+
+@[simp]
+theorem isSubgraph_eq_le : (Digraph.IsSubgraph : Digraph V → Digraph V → Prop) = (· ≤ ·) := rfl
+
+/-- The supremum of two digraphs `x ⊔ y` has edges where either `x` or `y` have edges. -/
+instance : Sup (Digraph V) where
+  sup x y := { Adj := x.Adj ⊔ y.Adj }
+
+@[simp]
+theorem sup_adj (x y : Digraph V) (v w : V) : (x ⊔ y).Adj v w ↔ x.Adj v w ∨ y.Adj v w := Iff.rfl
+
+/-- The infimum of two digraphs `x ⊓ y` has edges where both `x` and `y` have edges. -/
+instance : Inf (Digraph V) where
+  inf x y := { Adj := x.Adj ⊓ y.Adj }
+
+@[simp]
+theorem inf_adj (x y : Digraph V) (v w : V) : (x ⊓ y).Adj v w ↔ x.Adj v w ∧ y.Adj v w := Iff.rfl
+
+/-- We define `Gᶜ` to be the `Digraph V` such that no two adjacent vertices in `G`
+are adjacent in the complement, and every nonadjacent pair of vertices is adjacent. -/
+instance hasCompl : HasCompl (Digraph V) where
+  compl G := { Adj := fun v w ↦ ¬G.Adj v w }
+
+@[simp] theorem compl_adj (G : Digraph V) (v w : V) : Gᶜ.Adj v w ↔ ¬G.Adj v w := Iff.rfl
+
+/-- The difference of two digraphs `x \ y` has the edges of `x` with the edges of `y` removed. -/
+instance sdiff : SDiff (Digraph V) where
+  sdiff x y := { Adj := x.Adj \ y.Adj }
+
+@[simp]
+theorem sdiff_adj (x y : Digraph V) (v w : V) : (x \ y).Adj v w ↔ x.Adj v w ∧ ¬y.Adj v w := Iff.rfl
+
+instance supSet : SupSet (Digraph V) where
+  sSup s := { Adj := fun a b ↦ ∃ G ∈ s, Adj G a b }
+
+instance infSet : InfSet (Digraph V) where
+  sInf s := { Adj := fun a b ↦ (∀ ⦃G⦄, G ∈ s → Adj G a b) }
+
+@[simp]
+theorem sSup_adj {s : Set (Digraph V)} {a b : V} : (sSup s).Adj a b ↔ ∃ G ∈ s, Adj G a b := Iff.rfl
```

Comment:
Seems weird to me that `sSup_adj` lemma (explicitly) takes `{a b : V}` while the others don't
```suggestion
theorem sSup_adj {s : Set (Digraph V)} : (sSup s).Adj a b ↔ ∃ G ∈ s, Adj G a b := Iff.rfl
```
Or adding the parameter to the other lemmas